src/ray/core_worker/task_manager.cc

```diff
@@ -1018,6 +1031,24 @@ void TaskManager::MarkTaskReturnObjectsFailed(
       }
     }
   }
+  // If it was a streaming generator, try failing all the return object refs.
+  // In a normal time, it is no-op because the object ref values are already
+  // written, and Ray doesn't allow to overwrite values for the object ref.
+  // It is only useful when lineage reconstruction retry is failed. In this
+  // case, all these objects are lost from the plasma store, so we
+  // can overwrite them. See the test test_dynamic_generator_reconstruction_fails
+  // for more details.
+  if (spec.IsStreamingGenerator()) {
```

Comment:
Could you walk me through this offline.

I'm also wondering how you are handling the case where the first execution succeeded but with application level exceptions and the lineage reconstruction creates different number of objects.

Reply:
We don't handle that case. This assumes the task is deterministic, otherwise undefined behavior (same as regular reconstruction).

so the scenario you mentioned shouldn't happen (it means it is underministic).

Reply:
maybe we should have a stronger guarantee here. Like if it is undeterminstic, it should never hang but fail

Reply:
We can discuss all edge cases tmrw